File: testdata/programs/bench_sort.py
import time

class Item:
    def __init__(self, k: int, v: int):
        self.key = k
        self.val = v

def qsort(arr, lo, hi):
    if lo >= hi:
        return
    pivot = arr[lo + (hi - lo) // 2].key
    i, j = lo, hi
    while i <= j:
        while arr[i].key < pivot:
            i += 1
        while arr[j].key > pivot:
            j -= 1
        if i <= j:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
            j -= 1
    qsort(arr, lo, j)
    qsort(arr, i, hi)

started = time.perf_counter()

N = 5000
items = [Item(N - i + (i % 17) * 3, i) for i in range(N)]
build_ms = (time.perf_counter() - started) * 1000

t1 = time.perf_counter()
qsort(items, 0, N - 1)
sort_ms = (time.perf_counter() - t1) * 1000

checksum = sum(item.key for item in items)

ended = time.perf_counter()

print("python sort benchmark")
print(f"checksum={checksum}")
print(f"first_key={items[0].key}")
print(f"last_key={items[-1].key}")
print(f"build_ms={build_ms:.3f}")
print(f"sort_ms={sort_ms:.3f}")
print(f"elapsed_ms={(ended - started) * 1000:.3f}")

File: testdata/programs/bench_sort.lua
local function qsort(arr, lo, hi)
  if lo >= hi then return end
  local mid = lo + math.floor((hi - lo) / 2)
  local pivot = arr[mid].key
  local i, j = lo, hi
  while i <= j do
    while arr[i].key < pivot do i = i + 1 end
    while arr[j].key > pivot do j = j - 1 end
    if i <= j then
      local tmp = arr[i]; arr[i] = arr[j]; arr[j] = tmp
      i = i + 1; j = j - 1
    end
  end
  qsort(arr, lo, j)
  qsort(arr, i, hi)
end
local started = os.clock()
local N = 5000
local items = {}
for i = 0, N - 1 do items[i] = { key = N - i + (i % 17) * 3, val = i } end
local t1 = os.clock()
qsort(items, 0, N - 1)
local checksum = 0
for i = 0, N - 1 do checksum = checksum + items[i].key end
local ended = os.clock()
print("gopher-lua sort benchmark")
print("checksum=" .. checksum)
print("first_key=" .. items[0].key)
print("last_key=" .. items[N - 1].key)
print("elapsed_ms=" .. ((ended - started) * 1000.0))
